src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SslStreamSecurityUpgradeProvider.cs

```diff
@@ -566,16 +414,210 @@ internal override void Close(TimeSpan timeout)
+                    protectionLevel = SocketProtectionLevel.Tls10;
+                if ((_parent.SslProtocols & SslProtocols.Tls11) != SslProtocols.None)
+                    protectionLevel = SocketProtectionLevel.Tls11;
+                if ((_parent.SslProtocols & SslProtocols.Tls12) != SslProtocols.None)
+                    protectionLevel = SocketProtectionLevel.Tls12;
+
+                // With SslStream, the hostname provided in the server certificate is provided to the client and verified in the callback.
+                // With UWP StreamSocket, the hostname needs to be provided to the call to UpgradeToSslAsync. The code to fetch the identity
+                // lives in the callback for CoreClr but needs to be pulled into this method for UWP.
+                EndpointAddress remoteAddress = RemoteAddress;
+                if (remoteAddress.Identity == null && remoteAddress.Uri != Via)
+                {
+                    remoteAddress = new EndpointAddress(Via);
+                }
+                EndpointIdentity identity;
+                if (!_parent.IdentityVerifier.TryGetIdentity(remoteAddress, out identity))
+                {
+                    SecurityTraceRecordHelper.TraceIdentityVerificationFailure(identity, null, GetType());
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(
+                        new MessageSecurityException(SR.Format(SR.IdentityCheckFailedForOutgoingMessage, identity,
+                            remoteAddress)));
+                }
+                Contract.Assert(identity.IdentityClaim.ClaimType == ClaimTypes.Dns);
+                string dnsHostName = identity.IdentityClaim.Resource as string;
+
+                // This is the actual call to negotiate an SSL connection
+                await rtStreamSocket.UpgradeToSslAsync(protectionLevel, new HostName(dnsHostName)).AsTask();
+
+                // Verify that we didn't negotiate a protocol lower than the binding configuration specified. No need to check Tls12 
+                // as it will only be negotiated if Tls12 was actually specified. 
+                var negotiatedProtectionLevel = rtStreamSocket.Information.ProtectionLevel;
+                if ((negotiatedProtectionLevel == SocketProtectionLevel.Tls11 && (_parent.SslProtocols & SslProtocols.Tls11) == SslProtocols.None) ||
+                    (negotiatedProtectionLevel == SocketProtectionLevel.Tls10 && (_parent.SslProtocols & SslProtocols.Tls) == SslProtocols.None))
+                {
+                    // Need to dispose StreamSocket as normally SslStream wouldn't end up in a usable state in this situation. As
+                    // post-upgrade validation is required in UWP, the connection needs to be Dispose'd to ensure it isn't used.
+                    rtStreamSocket.Dispose();
+                    throw new SecurityNegotiationException(negotiatedProtectionLevel.ToString());
```

Comment:
Should we introduce a new resource here and produce a more meaningful message?  I know this is UWP and we won't see the resource content, but the resource name might offer better clues to the failure.